modin/core/execution/dask/implementations/pandas_on_dask/partitioning/partition.py

```diff
@@ -64,9 +62,9 @@ def get(self):
         """
         self.drain_call_queue()
         # blocking operation
-        if isinstance(self.future, pandas.DataFrame):
+        if not isinstance(self.future, Future):
```

Comment:
Why do we need to change this?

Reply:
Thus, the code looks clearer, we materialize when the object can be materialized.

Reply:
A partition should hold a pandas DataFrame. We may come to undefined behavior or errors in this case.

Reply:
[Some code](https://github.com/modin-project/modin/pull/3991/files#diff-348036bdb519cf0606cdd3a08df8e5483c16fb9c39fe9f3d30983758fa391551R126) is written as if there could also be an object of `pandas.series` type.

Reply:
Each partition holds a pandas DataFrame, but not a pandas Series. That is what we require and say in docs. So I am in favor of checking for a pandas DataFrame only, to be strongly restrict.

Reply:
I think we are talking about different things. Checking for which objects can be stored in a partition should be as early as possible, for example in the constructor. Here it is the check whether it is possible to call `materialize` function on the stored object.

Reply:
You can't guarantee that a future in the constructor is exactly a pandas DataFrame. Otherwise, you'll break up asynchronous execution. Here we should check whether the object is a future to materialize or a pandas DataFrame to just return. We should probably add a check that materialized data is a pandas.DataFrame, and also, if a remote call is supposed to return a non pandas.DataFrame, we should execute the task via DaskWrapper in other places.

Reply:
Why does `isinstance` function break asynchronous execution? Does it look at what lies inside the future?

Reply:
How can a materialized object appear in a partition? At what point in time? I thought that only through the constructor. 

If the object is `future` type, then it is obviously not necessary to check what is inside until the moment of the necessary materialization.

Reply:
> How can a materialized object appear in a partition? At what point in time? I thought that only through the constructor.

Only a future (for Dask) and an object ref (for Ray) are allowed in the Partition constructor.

> If the object is future type, then it is obviously not necessary to check what is inside until the moment of the necessary materialization.

That is true.

Reply:
It turns out that this check is not needed, since at the time of the `isinstance` call, there is always an object of `Future` type. Let's remove it.

As indirect evidence that this check is not needed, we can look at [the implementation](https://github.com/modin-project/modin/blob/b7027597985a442734a57949d72e3fa0625b4d43/modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py#L60) of this function for ray.

Reply:
1. We should add a check for a future passed in the constructor to detect if it is really Dask.Future (as for Ray).
2. I still tend to add a check in get if the materialized data is a pandas DataFrame.
3. We should remove isinstance for a pandas Series in to_pandas.

Reply:
https://github.com/modin-project/modin/pull/4207

Reply:
2. I see, to_numpy is presumably the only case when we have a NumPy array in a partition. Okay, let's leave it as is but we should think of this.
3. We can remove isinstance for a pandas Series in to_pandas since we don't add that check in get.

Reply:
3. This should not be done, because after calling some function on the `DataFrame` object, an object of `Series` type may be obtained, which will be stored in the partition (in that case, `get` function will return `Series`).